lib/eosknowledgeprivate/ekn-media-bin.c

```diff
@@ -637,26 +601,208 @@ ekn_media_bin_video_sink_new (EknMediaBin *self)
-      GST_WARNING ("Falling back to fakesink");
+      GST_WARNING ("Could not get video widget from gtkglsink/gtksink, falling back to fakesink");
 
       g_clear_object (&video_sink);
       video_sink = gst_element_factory_make ("fakesink", "EknMediaBinFakeSink");
 
-      gtk_container_add (GTK_CONTAINER (priv->overlay), img);
-      gtk_widget_show (img);
+      gtk_image_set_from_icon_name (GTK_IMAGE (priv->image),
+                                    "image-missing",
+                                    GTK_ICON_SIZE_DIALOG);
+      gtk_stack_set_visible_child (priv->stack, priv->image);
 
       /* FIXME: the overlay does not get motion and press events with this code path */
     }
 
-  return video_sink;
+  /* Setup playbin video sink */
+  if (video_sink)
+    {
+      g_object_set (priv->play, "video-sink", video_sink, NULL);
+      priv->video_sink = g_object_ref (video_sink);
+    }
+}
+
+static gboolean
+quit_when_idle (gpointer loop)
+{
+  g_main_loop_quit (loop);
+
+  return G_SOURCE_REMOVE;
+}
+
+static void
+check_for_draw (GdkEvent *event, gpointer loop)
+{
+  if (event->type == GDK_EXPOSE)
+    {
+      g_idle_add (quit_when_idle, loop);
+      gdk_event_handler_set ((GdkEventFunc) gtk_main_do_event, NULL, NULL);
+    }
+
+  gtk_main_do_event (event);
+}
+
+/* Based on gtk-reftest.c  */
+static void
+widget_drawing_sync (GtkWidget *widget)
+{
+  GdkWindow *window = gtk_widget_get_window (widget);
+  GMainLoop *loop;
+
+  loop = g_main_loop_new (NULL, FALSE);
+  /* We wait until the widget is drawn for the first time.
+   * We can not wait for a GtkWidget::draw event, because that might not
+   * happen if the window is fully obscured by windowed child widgets.
+   * Alternatively, we could wait for an expose event on widget's window.
+   * Both of these are rather hairy, not sure what's best. */
+  gdk_event_handler_set (check_for_draw, loop, NULL);
+  g_main_loop_run (loop);
+
+  /* give the WM/server some time to sync. */
+  gdk_display_sync (gdk_window_get_display (window));
+  g_timeout_add (12, quit_when_idle, loop);
+  g_main_loop_run (loop);
+
+  g_main_loop_unref (loop);
+}
+
+static void
+ekn_media_bin_fullscreen_apply (EknMediaBin *self, gboolean fullscreen)
+{
+  EknMediaBinPrivate *priv = EMB_PRIVATE (self);
+  gint64 position = -1;
+
+  /*
+   * To avoid flickering, this will make the widget pack a image with the last
+   * frame in the container before reparenting the video widget in the
+   * fullscreen window
+   */
+  if (!priv->tmp_image)
+    {
+      GdkPixbuf *pixbuf = ekn_media_bin_video_pixbuf_new (self);
+      priv->tmp_image = gtk_image_new_from_pixbuf (pixbuf);
+      g_object_set (priv->tmp_image, "expand", TRUE, NULL);
+      g_object_unref (pixbuf);
+    }
+
+  /*
+   * FIXME: GtkGstGLWidget does not support reparenting to a different toplevel 
+   * because the gl context is different and the pipeline does not know it
+   * changes, so as a temporal workaround we simply reconstruct the whole 
```

Comment:
Unless you mean to say that this workaround is of this material world, and not of the kingdom of heaven :-)